Feature/ci docker in docker

rmelotte opened the pull request:
Shell executors have a number of limitations:
- they may keep build results from one run to the other.
- they may not be able to remove previous built results made by
  root (a docker container for instance).
- when they use artifacts from a docker executor, they may have
  problems with file ownership and issue warnings like the following:
' WARNING: build/: lchown build/: operation not permitted (suppressing repeats) '
- the runners using shell executors are harder to reproduce on other
  hosts.

For those reasons, avoid using shell executors at all, and use docker
executors only.

One of the big required changes is that the build directory (when using the docker builder) is now inside the repository itself (under "build").

Also, since running the tests requires to use docker _inside_ docker, the docker-dind image and privileged runners are used.

arnout committed:
gitlab-ci.yml: do build step directly in builder image

Instead of pulling the docker images in the build step, use a normal
runner and build directly in the prplmesh-builder image pulled from
dockerhub. Override the entrypoint since the default one is to run
maptools.py. Note that pulling the runner image still needs to be done
in the run-tests step.

Note that we now use the clang-format.sh script from the top-level
directory instead of the one from tools/docker. The latter one just
calls the one in the top-level directory inside the prplmesh-builder
container.

The build results will still be in the same location and passed on to
the next stage via an artifact.

Since the build step now uses the normal docker-based build process, it
is not needed to specify tags for it.

The before and after scripts are only needed for the run-tests stage, so
move them there.

Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>
Signed-off-by: Raphaël Mélotte <raphael.melotte@mind.be>

rmelotte committed:
tools: docker: remove the jq dependency

jq was only used in one place to get some parts of the configuration
of the docker network.

Since it's possible to get this information from docker directly using
the "--filter" option, use that instead. Also since jq is not a
requirement anymore, do not check for it's availability in CI.

This makes it possible to use run.sh on hosts or docker images that do
not contain jq.

Signed-off-by: Raphaël Mélotte <raphael.melotte@mind.be>

rmelotte committed:
tools: make topdir the top-level directory of the repository

Using the parent folder of the repository as the top directory has a few
disadvantages:
- you can't have multiple worktree in a single folder, you need to add
an additional subfolder each time.
- the folder name of the repository is hardcoded to 'prplMesh'.
- the CI clones the repository in the working directory, not in a
'prplMesh' subfolder.
- some CI don't support artifacts outside of the repository.

This commit changes the 'topdir' variable to be the top-level
directory (i.e. the result of `git rev-parse --show-toplevel` *if*
you're using git).

The variable is also renamed to 'rootdir', to try to avoid potential
confusion for other developers who might not have noticed this change.

The following command was run at the top-level directory to generate
the first part of this commit:

find . -name '*.sh' \
-exec sed -i 's:\${topdir}/prplMesh/tools/docker/functions.sh:\${rootdir}/tools/docker/functions.sh:g' {} \; \
-exec sed -i 's:topdir="\${scriptdir%/\*:rootdir="\${scriptdir%:g' {} \; \
-exec sed -i 's/topdir/rootdir/g' {} \;

After that, a few manual fixes are needed:

.gitlab-ci.yml: the build artifacts is now in the current directory,
and the entrypoint of the run script has moved.

.travis.yml: the entrypoint of the run script has moved.

tools/commands/build.py: the default build directory should now be
"build".

tools/docker/builder/Dockerfile: the entrypoint has moved.

tools/docker/functions.sh: the rootdir variable (which was previously
the sourcesdir variable) has to be removed (the sourcesdir is now
rootdir).

tools/docker/tests/test_flows.sh: the send_CAPI_command.py is no
longer in 'prplMesh'.

.vscode/launch.json: the install directory has moved.

Since this change requires rebuilding the docker images and may break
other setups, add a "prplMesh -> ." symlink at the root of the repository
for now.

Signed-off-by: Raphaël Mélotte <raphael.melotte@mind.be>

rmelotte committed:
.gitlab-ci.yml: add the docker tags when building prplMesh

We also want the build-in-docker job to be run with a docker executor
since we need to have control over the clang-format version, so add
the "docker" tag.

Remove the "privileged" tag from the run-tests job, since using both
the tags "docker" and "docker-build" now means that docker-in-docker
is available (privileged was previously used for that, but we might
also have shell runners with the privileged tag).

Signed-off-by: Raphaël Mélotte <raphael.melotte@mind.be>

rmelotte committed:
tools: stop.sh: be less verbose

Echoing when there is nothing to do makes it harder to quickly
understand why a CI job is failing.

When there is nothing to do, only exit.

Also, as the stop script may be run as the very last job of the CI, it
may try to stop or remove containers which were already stopped as
part of a successful test. In such situations, we don't want the error
messages to be displayed. For this reason, redirect stderr to "/dev/null".

When the containers are stopped successfully on the other hand, each
container name will be output. Those names are not useful to have in
the CI logs either, so redirect stdout to "/dev/null" as well.

Signed-off-by: Raphaël Mélotte <raphael.melotte@mind.be>

rmelotte committed:
tools: send_CAPI_command.py: add a small configurable timeout

When the socket cannot be reached for some reason (the controller is
not running anymore for example) it takes a long time (>10 minutes)
for the current CI tests to fail. This is because the socket doesn't
have a timeout, so the timeout only comes from the system network
stack after some time.

Add a timeout argument to the constructor with a default value of 10.

With this change, the test should fail far more quickly if the
controller is not running anymore.

Signed-off-by: Raphaël Mélotte <raphael.melotte@mind.be>

rmelotte committed:
.gitlab-ci.yml: do not detach the tlvf_test container

The "-d" option was mistakenly added to the run arguments. Setting
"-d" immediately detaches the container, so the result of the test is
never checked.

Remove the "-d" option, to check the test result.

Signed-off-by: Raphaël Mélotte <raphael.melotte@mind.be>

rmelotte committed:
tools: run.sh: remove the "tty" and "interactive" options

With commit df673fb0ba1a42c90c24a14cf2d2de2d3741ba3a (tools: docker:
remove the "tty" and "interactive" options), the "--tty" and
"--interactive" options were removed to be able to be able to run the
tests non-interactively.

However the run.sh script slipped through and is still using those
options, so remove them.

Signed-off-by: Raphaël Mélotte <raphael.melotte@mind.be>

rmelotte committed:
.gitlab-ci.yml: upload prplMesh logs as artifacts

Uploading the logs as artifacts allows debugging tests after the CI
ran.

Mount the logs directory of each containers as a volume so that they
can be retrieved from the host, and upload it as an artifact. The
container name is used to mount the volume in logs/$name.

"when: always" is used in the CI configuration to upload the artifact
no matter if the test succeeded or not.

Signed-off-by: Raphaël Mélotte <raphael.melotte@mind.be>

rmelotte committed:
.gitlab-ci.yml: use a docker executor to run tests

Shell executors have a number of limitations:
- they may keep build results from one run to the other.
- they may not be able to remove previous built results made by
  root (a docker container for instance).
- when they use artifacts from a docker executor, they may have
  problems with file ownership and issue warnings like the following:
' WARNING: build/: lchown build/: operation not permitted (suppressing repeats) '
- the runners using shell executors are harder to reproduce on other
  hosts.

For those reasons, avoid using shell executors at all, and use docker
executors only.

Since the test stage needs to run some docker containers, we need to
be able to run docker containers inside the CI container. For that, we
need a privileged runner, so add the "privileged" tag.

To run docker inside docker, the following is possible:
- use docker-dind images and privileged runners.
- bind-mount the docker socket with a volume.

Because the second option requires unique container and network names
for concurrent jobs and makes it more difficult to mount volumes from
the source repository, we chose the first option (docker-dind
images). One of the downsides of this approach is that layer caching
may not work as expected (the images we build here are relatively
small, so we can live with this downside for now).

We need some more packages than what is available in the docker-dind
image, so add a Dockerfile to avoid having to install the same
packages at each run.
Note that the Dockerfile is not currently built as part of the CI, the
image needs to be present in the CI registry. It currently uses a
specific version of the docker-dind image, because the newer ones may
still have issues with regard to TLS and Gitlab runners.

The following packages are currently needed:
- python3 is required at least to run send_CAPI_command.py.
- grep is used in test_flows.sh (busybox's version doesn't have '-P')
- the coreutils version of 'tr' is used in test_flows.sh (busybox's
  version doesn't have '--delete').

We cannot currently use the docker-dind image as a service, because
the interfaces created by "docker network create" are only visible in
the dind container (and we currently rely on them to be able to send
CAPI commands).

Also, since we can now be sure python3 is installed, remove the
command check.

Signed-off-by: Raphaël Mélotte <raphael.melotte@mind.be>

Testing entrypoints

Signed-off-by: Raphaël Mélotte <raphael.melotte@mind.be>

rmelotte committed:
.gitlab-ci.yml: switch to cmake and make the build type a variable

Now that we know the docker-based builds work, we can switch to using
cmake and Ninja, like we instruct in the README.

While we're at it, make the build type a variable, so that it can be
overridden when manually triggering a build.

Signed-off-by: Raphaël Mélotte <raphael.melotte@mind.be>

arnout commented on .gitlab-ci.yml:
```diff
@@ -14,14 +14,14 @@ build-in-docker:
     # Try to pull from dockerhub, build the container images if that fails
     - ./image-pull.sh || ./image-build.sh
-    - ./run.sh -f -d --entrypoint "$PWD/../../../build/install/bin/tests/tlvf_test"
+    - ./run.sh -f -d --entrypoint "$PWD/../../build/install/bin/tests/tlvf_test"
```
The same change needs to be done in `.travis-ci.yml`.

rmelotte replied:
Done.

arnout commented on tools/docker/builder/Dockerfile:
```diff
@@ -34,4 +34,4 @@ RUN apt-get update && apt-get install -y \
     && rm -rf /var/lib/apt/lists/*
 
-ENTRYPOINT ["./prplMesh/tools/maptools.py", "build"]
+ENTRYPOINT ["./tools/maptools.py", "build"]
```
This is really annoying - it requires a rebuild and push of the docker containers or things will break.

I think this can be solved by adding a symlink `prplMesh -> .` to the rootdir. So please do that as well in the same commit.

rmelotte replied:
Done.

arnout commented on .gitlab-ci.yml:
```diff
@@ -15,6 +15,10 @@ build-in-docker:
       - build
+  tags:
+    - docker
+    - specific
```
NACK this. We can just configure the project on gitlab.com to not use the shared runners. But I don't think we need to - I believe the "private" runners are used in preference to shared runners, so the shared runners would only be used when the private one is already overloaded.

rmelotte replied:
Removed the "specific" tag.

arnout commented on .gitlab-ci.yml:
```diff
@@ -21,7 +21,7 @@ run-tests:
     # Try to pull from dockerhub, build the container images if that fails
     - ./image-pull.sh || ./image-build.sh
-    - ./run.sh -f -d --entrypoint "$PWD/../../build/install/bin/tests/tlvf_test"
+    - ./run.sh -f --entrypoint "$PWD/../../build/install/bin/tests/tlvf_test"
```
A bit useless to do this since in the end we want to move this to a separate stage anyway, but OK.

arnout commented on tools/docker/tests/test_flows.sh:
```diff
@@ -35,7 +35,7 @@ send_bml_command() {
 send_CAPI_command() {
     # send a CAPI command to a container.
-    ${rootdir}/tools/docker/tests/send_CAPI_command.py $(container_ip "$1") $(container_CAPI_port "$1") "$2"
+    ${rootdir}/tools/docker/tests/send_CAPI_command.py $(container_ip "$1") $(container_CAPI_port "$1") "$2" "$verbose_option"
```
I think it's more appropriate to just put `dbg` in front instead.

rmelotte replied:
Done, I dropped the entire commit and used dbg instead.

arnout commented on tools/docker/run.sh:
```diff
@@ -105,6 +105,7 @@ main() {
                 --expose ${PORT}
                 -v ${installdir}:${installdir}
                 -v ${rootdir}:${rootdir}
+		-v ${rootdir}/logs/${NAME}:/tmp/${SUDO_USER:-${USER}}/beerocks/logs
```
Indentation is wrong here.

rmelotte replied:
That's only Github playing tricks on us. When you open the file, the indentation looks correct.

arnout commented on .gitlab-ci.yml:
```diff
@@ -33,6 +33,11 @@ run-tests:
+  artifacts:
+    paths:
+      - logs
+    when: always
```
Maybe `on_failure` would be more appropriate?

rmelotte replied:
It doesn't hurt to keep them, they could be used as a reference for someone wanting to check what the logs look like in a normal flow. They are relatively small and will be deleted after some time anyway.
What do you think?

arnout commented on .gitlab-ci.yml:
```diff
@@ -1,31 +1,49 @@
+
+  artifacts:
+    paths:
+      - build
```
Do we really need to save the entire build directory? Isn't build/install sufficient?

rmelotte replied:
Done.

arnout commented on .gitlab-ci.yml:
```diff
@@ -1,31 +1,49 @@
+
+  before_script:
+    # start the docker daemon and give it some time to start:
+    - dockerd-entrypoint.sh > /dev/null 2>&1 & sleep 2
```
Maybe better do an appropriate wait loop, e.g. `for i in $(seq 1 10); do docker info -f '.' && break; sleep .5; false; done`. Note that the `false` makes sure we error out if docker is not alive.

However, why can't we just set this as the ENTRYPOINT in our tests-runner image? Or specify it explicitly with

```
image:
  name: $CI_REGISTRY_IMAGE/tests-runner
  entrypoint: dockerd-entrypoint.sh
```

rmelotte replied:
We cannot use the entrypoint as-is because gitlab-ci needs the entrypoint to either be empty, or to be a shell. For reference, you can see _part_ of what gitlab-ci is doing with the entrypoint here: https://gitlab.com/prpl-foundation/prplMesh/-/jobs/378489864
In a few words: it first exec the entrypoint, then gives it a script that, among other things, tries to exec another shell. After that (not visible in this log), the content of the "script" part of `.gitlab-ci.yml` is passed through stdin (along with a bunch of added commands and variables; `set -eo pipefail` is one of them).

I created a new entrypoint that uses the loop you describe, then runs the gitlab-ci script at the end (or fail if dockerd could not be started).

arnout reviewed (changes requested):
I'd like to try to pass both gitlab-ci and Travis before merging.

rmelotte committed:
tools: tests_flows.sh: display the CAPI commands details when verbose

It can be useful to have more details about commands being sent by
send_CAPI_command.py.

Add a debug message when running send_CAPI_command, so that the
command and it's arguments are displayed if the verbose flag was
provided.

Signed-off-by: Raphaël Mélotte <raphael.melotte@mind.be>

rmelotte committed:
.gitlab-ci.yml: only keep the install dir as an artifact

We don't need the whole build directory to be kept as an artifact,
only keeping build/install is sufficient.

Signed-off-by: Raphaël Mélotte <raphael.melotte@mind.be>

arnout reviewed (approved):
There are still a few references to ../build left:

- .vscode/launch.json
- README.md

But that can be fixed later.

There was really no need to do the last commit as a separate commit, it could be done immediately like that in 22540d55. But that's a nitpick.

tomereli commented on tools/docker/tests/test_flows.sh:
```diff
@@ -35,7 +35,7 @@ send_bml_command() {
 send_CAPI_command() {
     # send a CAPI command to a container.
-    ${topdir}/prplMesh/tools/docker/tests/send_CAPI_command.py $(container_ip "$1") $(container_CAPI_port "$1") "$2"
+    ${rootdir}/prplMesh/tools/docker/tests/send_CAPI_command.py $(container_ip "$1") $(container_CAPI_port "$1") "$2"
```
How can this work? shouldn't this be `${rootdir}/tools/docker/tests/send_CAPI_command.py $(container_ip "$1") $(container_CAPI_port "$1") "$2"` ?

rmelotte replied:
This commit was generated using the command in the commit message. The next commit fixes this manually.
It would have been possible to include this particular change in the first (generated) commit, but other manual fixes are needed, so the 2 commits _have_ to be applied together anyway.

tomereli replied:
I now see that e74d848c7e71dc3bba8275f5a3db2c1436885cc3 fixes that but IMHO it should be squashed to the commit before that, otherwise this commit "breaks" `build.sh`, doesn't it? In that case I think it would be better to squash them so build will not break.

rmelotte replied:
2c435b4a indeed does break things, it has to be applied with the next one (e74d848c).
I remember @arnout mentioned he prefers to have "generated" commits separate from their manual fixes, but I don't know if it applies in this case.

rmelotte replied:
I asked @arnout, the reason to have them split at first is to make it easier to review.
Since you both already did, I'll squash them now.

tomereli commented on tools/docker/functions.sh:
```diff
@@ -5,8 +5,8 @@
-installdir="${topdir}/build/install"
-sourcesdir="${topdir}/prplMesh"
+installdir="${rootdir}/build/install"
+rootdir="${rootdir}/prplMesh"
```
Should be `rootdir="${scriptdir%/*/*}"` like in other places, shouldn't it?

tomereli reviewed (commented):
Commit 2c435b4af4a12dcc88c23a5a33a16fbc93958455 mentions that some manual fixes are needed after the mechanical replace of `topdir` to `rootdir` but I didn't see these fixed in followup commits, so I think these are real bugs - can you please double check?
